werkzeug/test.py

```diff
@@ -339,6 +345,19 @@ def __init__(self, path='/', base_url=None, query_string=None,
                     else:
                         self.form.setlistdefault(key).append(value)
 
+        if json:
+            if not isinstance(json, dict):
+                raise TypeError('json parameter must be of type dict')
+
+            json = _json.dumps(json)
+            json = json.encode(self.charset)
```

Comment:
I have a suspicion that this fails with non-ASCII data, because `json.dumps` returns unicode strings in Python 3 and bytestrings under Python 2.